### ISSUE TITLE
Add authentication to MongoDB server

### ISSUE THREAD

**Issue by Mctalian** [YashdalfTheGray/flintandsteel#209]
In light of a [recent event](http://it.slashdot.org/story/15/12/17/0319203/over-650-tb-of-data-up-for-grabs-from-publicly-exposed-mongodb-database) (thanks, @LongArmMcGee, for the tip), I looked at our own setup. While we are not included in that 684.8 TB (because the port we use isn't open to external traffic), we aren't using authentication. So theoretically, anyone internal can bypass the site completely and write bogus data into the database, or do the famous "drop" of all our databases.

@LongArmMcGee has another issue coming which I will reference after it gets posted that I think needs attention before we focus on this. This could be a band-aid solution, but overall I think we need to do more.

**Comment by Mctalian** [comment]
Thought, if we do this, it should be closer to full-blown production. Unless we can find a way to keep authenticaion "off" for development, which I'm sure we can.

**Comment by Mctalian** [comment]
We should definitely look at #210 first.

**Comment by YashdalfTheGray** [comment]
Once I deliver the backend route changes, I was actually going to deliver some hardening from the client side so that's already on my todo list.

**Comment by YashdalfTheGray** [comment]
This will be the next priority after #222.

**Comment by YashdalfTheGray** [comment]
Here's what I'm thinking for this, create an account for our server and change `server/server.js` to ask for a password when `gulp start:[dev|prod]` is run so that the node code can authenticate with the mongo server. 

Thoughts?

**Comment by Mctalian** [comment]
Should be fine for now. I can't think of anything better other than storing the credentials in Environment Vars and pulling them on start for production only and don't lock down the dev database.

**Comment by davethenipper** [comment]
Sounds perfect.  Also sounds like it won't take too much time.

**Comment by davethenipper** [comment]
Yash, are you planning on working on this?  I'd like to get this guy assigned to someone XD

**Comment by YashdalfTheGray** [comment]
Yep, assigning to myself right now.

**Comment by YashdalfTheGray** [comment]
Alright, so after doing some research, I figured out that you add 

```yaml
security:
    authorization: enabled
```
to the mongo config file and that turns on authentication. But before you turn on authentication, you have to have users created on your database. We could write a script but then passwords would be published online and anyone could get access to it. I think once we start the prod server, we're just going to custom put security on it rather than put security on every instance of mongo ever. 

Besides that, in the hardening phase, I'm going to work on securing the REST API so that access from that side is authorized. 

Seeing all of this information, I'm going to close this issue and we'll just have to keep in mind to lock down the prod database when we start it.

**Comment by Mctalian** [comment]
If we need to remember something then I'd say keep this open.

**Comment by YashdalfTheGray** [comment]
We should add it to the trello faq or something but not here. We should close this issue.

**Comment by Mctalian** [comment]
I'm going to push back a little bit here. Since there's something that actually needs to be done here before we can claim MVP, I think this should remain open. If you want to close an issue, I'd say #210 is a better candidate. Someone needs to do this once we figure out a permanent production solution and I only visit the FAQ when I have a Q to A. Not when I'm looking for things that need to be completed

**Comment by YashdalfTheGray** [comment]
I agree with closing #210. I'll PM you with the rest of my concerns.
